src/RileyMaze/userFunction.js

```diff
@@ -0,0 +1,244 @@
+        if (this._model)
+            this._unbindLevelModel();
+
+        this._model = model;
+        this._notifyHandler = model.connect('notify', this._setCode.bind(this));
+        this._setCode();
+    }
+
+    _setCode() {
+        const {name, args, perLevel, modelProp} = USER_FUNCTIONS[this._codeview.userFunction];
+
+        let code;
+        if (perLevel)
+            code = this._model[modelProp];
+        else
+            code = this._global[modelProp];
+
+        this._codeview.text = `function ${name}(${args.join(', ')}) {
+${code}
+}`;
+    }
+
+    _updateCode(funcBody) {
+        const {perLevel, modelProp} = USER_FUNCTIONS[this._codeview.userFunction];
+        // Block the normal notify handler that updates the code view, since we
+        // are propagating updates from the codeview to the GUI.
+        if (perLevel) {
+            GObject.signal_handler_block(this._model, this._notifyHandler);
+            try {
+                this._model[modelProp] = funcBody;
+            } finally {
+                GObject.signal_handler_unblock(this._model, this._notifyHandler);
+            }
+        } else {
+            GObject.signal_handler_block(this._global, this._globalNotifyHandler);
+            try {
+                this._global[modelProp] = funcBody;
+            } finally {
+                GObject.signal_handler_unblock(this._global, this._globalNotifyHandler);
+            }
+        }
+    }
+
+    _compile() {
+        const {name, args, getScope} = USER_FUNCTIONS[this._codeview.userFunction];
+        const code = this._codeview.text;
+        const scope = getScope();
+        scope.riley.queue.length = 0;
+        let userFunction;
+
+        try {
+            // eslint-disable-next-line no-new-func
+            const factoryFunc = new Function('scope', `\
+with(scope){
+${code}
+; if (typeof ${name} !== 'undefined') return ${name};
+}`);
+            userFunction = factoryFunc(scope);
+        } catch (e) {
+            if (!(e instanceof SyntaxError || e instanceof ReferenceError))
+                throw e;
+            this._codeview.setCompileResultsFromException(e);
+            return;
+        }
+
+        if (!userFunction) {
+            // Function definition deleted, or otherwise not working
+            this._codeview.setCompileResults([{
+                start: {
+                    line: 1,
+                    column: 1,
+                },
+                end: {
+                    line: 1,
+                    column: 1,
+                },
+                message: `Expected a function named ${name}.`,
+                fixme: `function ${name}(${args.join('\n')}) {\n}\n`,
+            }]);
+            return;
+        }
+
+        try {
+            userFunction();
+        } catch (e) {
+            this._codeview.setCompileResultsFromException(e);
+            return;
+        }
+        // FIXME: ASK FOR BETTER MESSAGE?
+        if (scope.riley.queue.length !== 8) {
```

Comment:
We have to move this somehow into a validation function that's separate for each user function panel. Otherwise the level editor will check if the Riley queue is full. (You can see this by adding a space or changing something else in the level code view so that it recompiles — you get "Instructions must have 8 moves")